### ISSUE TITLE
Create an umbrella measurement for all SLIs

### ISSUE THREAD

**Issue by oxddr** [kubernetes/perf-tests#787]
Currently all SLIs are measured by a separated, dedicated measurement. Once they are stable there is no reason to run them separately. We should add an umbrella measurement so we can simplify test config. 

/area slo
/area clusterloader

**Comment by oxddr** [comment]
/assign @oxddr 

I'll take care of this early October. Free to take in the meantime.

**Comment by oxddr** [comment]
FYI: @wojtek-t @mm4tt this is good first issue for someone willing to learn about scalability tests and clusterloader.

**Comment by oxddr** [comment]
Here is more detailed design of this change:

1. Add `CombinedMeasurement` - type implementing `Measurement` interface (this is needed, so we can have a single Measurement type for SLO):
  - grouping together N other measurements
  - forwarding calls to all individual measurements 
  - aggregate returned errors in gather phase
  - up to investigation is how to specify configuration for each individual measurement. Ideally, this should be transparent to individual Measurements

2. Register instance of `CombinedMeasurement` with all existing SLI-related measurements as `KubernetesSLO`. Take a look at https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/testing/load/config.yaml#L64 for the reference (all measurements from "Starting measurements" section but TestMetrics)

3. Use `KubernetesSLO` in density and load test. Follow https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/docs/experiments.md to enable it gradually (disabling old measurement at the same time)

**Comment by oxddr** [comment]
@hemanthmalla please let me know if there is anything unclear

**Comment by oxddr** [comment]
@hemanthmalla how is it going with this issues? Did you have a chance to work on this one?

**Comment by vamossagar12** [comment]
@oxddr do you think we can apply the things done in the PR for this issue? https://github.com/kubernetes/perf-tests/issues/794

**Comment by oxddr** [comment]
Could you point me to a specific PR? I suspect we can apply at least some
pieces based on this issue description.

**Comment by vamossagar12** [comment]
Sure. here it is: https://github.com/kubernetes/perf-tests/pull/1075

**Comment by wojtek-t** [comment]
It can be a building block - I don't want users to even need to be aware of what SLOs we have. They should just say "measure me all SLOs" and don't specific specific one.

**Comment by vamossagar12** [comment]
Ok. I can take a stab at it once the PR I posted above gets merged. Can you plz add more details to this?

**Comment by oxddr** [comment]
I'll try to add more details today/tomorrow. However I am open to
implementation ideas as long as we would be able to instantiate all SLO
measurements with minimal number of config lines and configurations
options.

**Comment by vamossagar12** [comment]
hey @oxddr the PR that I talked about got merged a couple of months ago. Do you think we can get started with this?

**Comment by wojtek-t** [comment]
@vamossagar12 - yes, we can start with it now. The goal is to have a single SLI measurement that will initialize all our existing SLIs (it's fine to start with merging the first two, e.g. networking programming SLI and dns-programming SLI as a starting point)

**Comment by vamossagar12** [comment]
Thanks @wojtek-t , let me start by looking at the SLIs and come up with a small description of changes here  before I proceed with the implementation.

**Comment by vamossagar12** [comment]
hi @wojtek-t , I got the chance to look at the code today. My earlier thought was to reuse the measurement_executor that was developed : https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/measurement/measurement_executor.go 

But, looking at the comments here again, it makes sense to have a new measurement. 

So, in the config yamls, users don't need to specify any individual SLI specific like here:

https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/testing/load/config.yaml#L104-L115

Instead it, would be 1 block for SLIMeasurement(we can rename it, just an example), and specify any SLI specific config within that as a param.

The above configs for DNS lookup/network programmring can be merged into one config with something like:

```
  - Identifier: SLIMeasurement
    Method: SLIMeasurement
    Params:
      action: start
      dnsLookupCheckProbesReadyTimeout: {{$PROBE_MEASUREMENTS_CHECK_PROBES_READY_TIMEOUT}}
      dnsLookupReplicasPerProbe: {{AddInt 2 (DivideInt .Nodes 100)}}
      networkProgrammingLatencyEnabled: {{$PROMETHEUS_SCRAPE_KUBE_PROXY}}
```

Does it make sense? I can start making the code/config changes it it does... Plz let me know.

**Comment by wojtek-t** [comment]
Yes - this is exactly the goal. I would call it SLO, but that's a naming we can figure out later.

Ideally, our default will be smart enough that we won't need all those params, but that's again can be addressed independently.

**Comment by vamossagar12** [comment]
@wojtek-t @oxddr i have a question this. while going through the code, i found that there's already a measurement called probes which inhouses the dns lookup measurement and network latency. But, the difference is, it creates those measurements and then each one of them is executed as a separate measurement- which is what needs to be avoided.
In the new implementation, once the measurement has been created, it would query the prometheus server every configured time interval and get the stats. 
What is confusing me is that, in the case of probes measurement, prometheus measurement query for dns/latency is invoked only in the gather phase. Or am I missing something here?

**Comment by wojtek-t** [comment]
You should NOT touch any of the existing measurements at all.

What you should do is to create a NEW measurement, that for each action, it will call an appropriate action on all underlying measurements. We should probably start only with the APIResponsiveness, and add new measurements over time.

**Comment by vamossagar12** [comment]
Ok may be I wasn't clear enough but I didn't mean to touch any of the existing measurements. What I meant was, in the new measurement, do i invoke the prometheus query to get the network latency/dns lookup metrics repeatedly or do i do it only once in the gather phase.

**Comment by wojtek-t** [comment]
You rely on the underlying measurements and whatever they are currently doing.

**Comment by vamossagar12** [comment]
Yeah that's what i had thought of doing as well, but the thing is dns lookup and network latency measurements are being created within the probes measurement: 
https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/measurement/common/probes/probes.go#L64-L71

So, they don't have any explicit definitions like other measurements.

**Comment by wojtek-t** [comment]
This is only registration - it doesn't matter for the purpose of it.

**Comment by vamossagar12** [comment]
@wojtek-t , i looked at it further. One problem that I am currently facing is, how to invoke the relevant methods of underlying measurements from the uber SLI Measurement. One approach which I had followed in the implementation of Wrapper measurements is that I could find the relevant measurement using the measurementManager object being passed here:

https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/measurement/measurement_executor.go#L33

This is one step before the execution of any of the implementations of measurements so it worked fine. 

In this case, since the goal is to execute the individual measurements from SLI measurement, the problem is how to get access to the relevant instances of DNSLookup measurement or APIResponsiveness measurement from the CombinedMeasurement. The only method which lets me do that is in MeasurementManager ;

https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/measurement/manager.go#L92

but neither this method not measurementmanager is visible in the implementations of measurements. 


###########################

Looking at this scenario, 1 difference that I see between wrapper measurement and combined measurement is that, in the former, we have same method but different Instances of it, but in the latter, we have multiple methods i.e measurements, each having their own instances.

So, either I make changes to measurement_executor to support such kinds of measurements or I make the measurement_manager visible at an individual level- which i don't think is a good idea. 

Going by the idea of making changes to measurement_executor, maybe we add some kind of config to signal it as a combined measurement. What are the underneath measurements is something that needn't be defined by the end user but rather defined at the code level. Whenever such kind of combined measurement is encountered by the measurement_executor, it will get the relevant instances of all underneath measurements and pass it along to the instance of the CombinedMeasurement. We can add that in the config of the CombinedMeasurement. From there onwards, it is the duty of the combined measurement to invoke the relevant actions of each of the underlying measurements. WDYT?

**Comment by wojtek-t** [comment]
> In this case, since the goal is to execute the individual measurements from SLI measurement, the problem is how to get access to the relevant instances of DNSLookup measurement or APIResponsiveness measurement from the CombinedMeasurement. The only method which lets me do that is in MeasurementManager ;

You're trying to do that dynamically, which is not the goal. Just hardcode.

**Comment by vamossagar12** [comment]
> > In this case, since the goal is to execute the individual measurements from SLI measurement, the problem is how to get access to the relevant instances of DNSLookup measurement or APIResponsiveness measurement from the CombinedMeasurement. The only method which lets me do that is in MeasurementManager ;
> 
> You're trying to do that dynamically, which is not the goal. Just hardcode.

Ok. I can hardcode it, I don't think that's a problem. But my original problem is still the same, how do I get the measurementInstance object of the underlying measurements from the Combined SLI measurement. I don't see any other way of doing what you mentioned above:

**What you should do is to create a NEW measurement, that for each action, it will call an appropriate action on all underlying measurements. We should probably start only with the APIResponsiveness, and add new measurements over time.**

I apologise I am dragging the question too long, but my thought process is these measurements have already been defined somewhere else. So, I don't need to recreate them. That's why I am trying to see if I can get the instance of that object and call the underneath action. But that option isn't available at a measurement level. OR is there any other way of invoking the actions of 1 measurement from another measurement?

**Comment by wojtek-t** [comment]
You don't need to obtain anything - the new measurement can create its own instances of the measurements.

**Comment by wojtek-t** [comment]
/assign @vamossagar12 

The initial PR to address it it #1479

**Comment by vamossagar12** [comment]
@wojtek-t , i guess this issue can be closed as the PR has been merged.

**Comment by wojtek-t** [comment]
/remove-lifecycle stale

 I would like to push it a bit further:
- move networkProgramminLatency SLI to this measurement too
- see if we're ready to move api-call latencies to it too
